crates/ruff_linter/src/rules/refurb/rules/slice_to_remove_prefix_or_suffix.rs

```diff
@@ -0,0 +1,479 @@
+use crate::{checkers::ast::Checker, settings::types::PythonVersion};
+use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast as ast;
+use ruff_python_semantic::SemanticModel;
+use ruff_source_file::Locator;
+use ruff_text_size::Ranged;
+
+/// ## What it does
+/// Checks for the removal of a prefix or suffix from a string by assigning
+/// the string to a slice after checking `.startswith()` or `.endswith()`, respectively.
+///
+/// ## Why is this bad?
+/// The methods [`str.removeprefix`] and [`str.removesuffix`],
+/// introduced in Python 3.9, have the same behavior
+/// and are more readable and efficient.
+///
+/// ## Example
+/// ```python
+/// filename[:-4] if filename.endswith(".txt") else filename
+/// ```
+///
+/// ```python
+/// if text.startswith("pre"):
+///     text = text[3:]
+/// ```
+///
+/// Use instead:
+/// ```python
+/// filename = filename.removesuffix(".txt")
+/// ```
+///
+/// ```python
+/// text = text.removeprefix("pre")
+/// ```
+///
+/// [`str.removeprefix`]: https://docs.python.org/3/library/stdtypes.html#str.removeprefix
+/// [`str.removesuffix`]: https://docs.python.org/3/library/stdtypes.html#str.removesuffix
+#[violation]
+pub struct SliceToRemovePrefixOrSuffix {
+    string: String,
+    affix_kind: AffixKind,
+    stmt_or_expression: StmtOrExpr,
+}
+
+impl AlwaysFixableViolation for SliceToRemovePrefixOrSuffix {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        match self.affix_kind {
+            AffixKind::StartsWith => {
+                format!("Prefer `removeprefix` over conditionally replacing with slice.")
+            }
+            AffixKind::EndsWith => {
+                format!("Prefer `removesuffix` over conditionally replacing with slice.")
+            }
+        }
```

Comment:
```suggestion
        let replacement = self.affix_kind.replacement();
        format!("Prefer `{replacement}` over conditionally replacing with slice.")
```

Reply:
This one I kept as is, because of this suggestion from a previous PR: https://github.com/astral-sh/ruff/pull/12691#discussion_r1704822876

Reply:
Oh nice, thank you! I didn't consider the docs-generation impact of this :-)